### PR TITLE
[IMP] iot: Forward IoT logs to the server

### PR DESCRIPTION
**Pull request by lse-odoo**
Before this commit:
 IoT logs are kept in an IoT log file.
 Any time the support would need IoT log information, we are forced to ask the customer to send it to us as it requires to be in the LAN.

After this commit:
 Relevant* log lines will be automatically send out to the server using
 an HTTP route.
 This feature can be toggled within the Handlers list page

preview:

*: Relevant =
 - Any odoo logs (depending on the level set in the handlers list, see: https://github.com/odoo/odoo/pull/134174 )
 - Any other logs (werkzeug, python libraries, etc.) except /hw_proxy/hello

Note: IoT logs received by the server will ALWAYS be logged regardless of it level. So an Odoo server set in INFO which receive a DEBUG log from the IoT will log it in its log with the DEBUG level

Related enterprise PR: https://github.com/odoo/enterprise/pull/55055

opw-3696519

### CODE REVIEW

**Comment by lse-odoo** [comment]
Perspective of using other handler than HTTP handler as:
 - Calls are blocking: so the entire IoT wait until the POST request response is received !! Not great for IoT
 - It forces to use POST with raw parameters, using javascript is more secure (?)

SocketHandler could be a good candidate:
https://docs.python.org/3/library/logging.handlers.html#sockethandler
But I was not able to make it interact with an odoo.sh databases (I am not sure of the port to use), and the "blocking" aspect will likely also happen with this one.

Maybe better to have a more complicated queue system to send the logs with some delay but in order (the timestamp will be different from the server anyway and it's log time is given to the server).
Looks like logger have some built-in Queue related handler: https://docs.python.org/3/library/logging.handlers.html#queuehandler
Need to investigate more on how we could use it :thinking:

**Comment by lse-odoo** [comment]
TODO: add a flag on the handlers list page to allow to deactivate the feature

**Comment by lse-odoo** [comment]
@odoo/rd-security might I have your opinion on this PR please ?
Could you also have a look at the enterprise related PR: https://github.com/odoo/enterprise/pull/55055 ?

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:85]: @xmo-odoo any suggestions of format to use then ? maybe just regular POST ?

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:89]: @xmo-odoo can you reformulate please ? I am not sure to understand this point. Is it due to the previous except ?

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py]: I did attempt something that is quite light in the last push, let me know your thought on the matter.

However, I noticed that `get_data()` (on the enterprise PR) will have to wait for the full request content. Was your idea to iterate on the response even if we received partially or only to start the request early ?
The idea come from:
https://requests.readthedocs.io/en/latest/user/advanced/#chunk-encoded-requests

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:77]: @xmo-odoo ok, smart idea indeed

**Comment by lse-odoo** [comment]
@xmo-odoo 
Would you mind reformulating this message:
> Seems ok. I think the pseudo-binascii first line is unwarranted given the others don't bother with more complicated framing than newlines, and the first line already has a clear prefix.

I don't understand what you mean by that

**Comment by lse-odoo** [comment]
> > I don't understand what you mean by that
> 
> The entire payload is textual, except it uses 0x1f as line separator which doesn't seem warranted and will make tracing and debugging the payload a lot harder, as tools will either show a text view and ignore the US, printing a giant line with everything concatenated, or they'll switch to binary mode and refuse to show the payload or show a hex view or something.

Indeed. I did modify the separator with b'<log/>\n'. I also noticed that it could cause issue with the .rstrip() that was used due to:
https://stackoverflow.com/a/1038845/16356970
So I modified this part as well

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:121]: yes indeed. But I think it's better to call `self.toggle_active(False)` in the close() as it will itself call _close_content() (will also remove this method as it is called only once, I will merge it in the toggle_active for simplicity)

File: addons/hw_drivers/server_logger.py
Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:118]: I'm not even sure that this call is actually important. The thread is not run in daemon, so even if the Timeout is reached nothing should happen:
https://stackoverflow.com/a/24855374/16356970
Do you think it would be better to just use `.join()` ?

Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py]: @xmo-odoo would you have some news on this ? :)

File: addons/hw_drivers/server_logger.py
Comment (Ysoroko) [inline addons/hw_drivers/server_logger.py:104]: no need for "as _"

Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:104]: in theory not indeed, but if we don't the ci/style is not happy

File: addons/hw_drivers/server_logger.py
Comment (qle-odoo) [inline addons/hw_drivers/server_logger.py:19]: You instancy a constant to use it only one time

**Comment by lse-odoo** [comment]
@qle-odoo could you review it please ?

**Comment by qle-odoo** [comment]
@lse-odoo can you check your tests ?
ci/style

**Comment by qle-odoo** [comment]
@robodoo delegate+

File: addons/hw_drivers/server_logger.py
Comment (Ysoroko) [reply addons/hw_drivers/server_logger.py:108]: Probably because of the catching of the generic "Exception", right? You could avoid this by catching `requests.exceptions.RequestException` instead as the only thing which can throw an exception here is the odoo_session.post and RequestException is the parent class of all the exceptions coming from python requests (see [here](https://requests.readthedocs.io/en/latest/_modules/requests/exceptions/))

File: addons/hw_drivers/server_logger.py
Comment (Ysoroko) [inline addons/hw_drivers/server_logger.py:119]: tell me if I understand correctly: you try/except the full queue instead of using Queue.full() method to make it faster?

Comment (lse-odoo) [reply addons/hw_drivers/server_logger.py:119]: not really a question of speed, `put_nowait` can raise this exception if it is Full, so for reading  convenience it does make sense rather than a if not queue.full()
furthermore the documentation state, https://docs.python.org/3/library/queue.html#queue.Queue.full:
`if full() returns False it doesn’t guarantee that a subsequent call to put() will not block.`
even if here it is not blocking anyway.
I guess the final result would be the same, but I prefer this approach as it make more sense in my opinion

**Review by Ysoroko — APPROVED** [review]

**Comment by lse-odoo** [comment]
@robodoo r+

**Comment by lse-odoo** [comment]
robodoo r+

**Pull request closed without merging**
